open-hackathon/src/hackathon/docker/__init__.py

```diff
@@ -44,23 +60,29 @@ def containers_info(self, docker_host):
             raise
 
     def __get_available_host_port(self, port_bindings, port):
-        host_port = port + 10000
 
-        while host_port in port_bindings:
-            host_port += 1
-
-        if host_port >= 65535:
-            log.error("port used up on this host server")
-            raise Exception("no port available")
-
-        log.debug("host_port is %d " % host_port)
-        return host_port
+        self.lock.acquire()
+        try:
+            host_port = port + 10000
+            while host_port in port_bindings or host_port in OssDocker.host_ports:
+                host_port += 1
+            if host_port >= 65535:
+                log.error("port used up on this host server")
+                raise Exception("no port available")
+            if len(OssDocker.host_ports) >= 30:
+                self.__ports_cache()
+            OssDocker.host_ports.append(host_port)
+            log.debug("host_port is %d " % host_port)
+            return host_port
+        finally:
+            self.lock.release()
```

Comment:
When two threads visit static variable host_ports, if a thread run to line 70 and the other thread run to line 67, the second thread will have the same host_port, then will raise a conflict. So I use  synchronize method to avoid this case.